lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -0,0 +1,152 @@
+}
+Operation *InnerSymbolTable::lookup(StringAttr name) const {
+  return symbolTable.lookup(name);
+}
+
+//===----------------------------------------------------------------------===//
+// InnerSymbolTableCollection
+//===----------------------------------------------------------------------===//
+
+InnerSymbolTable &
+InnerSymbolTableCollection::getInnerSymbolTable(Operation *op) {
+  auto it = symbolTables.try_emplace(op, nullptr);
+  if (it.second)
+    it.first->second = ::std::make_unique<InnerSymbolTable>(op);
+  return *it.first->second;
+}
+
+void InnerSymbolTableCollection::constructTablesInParallelFor(
+    ArrayRef<Operation *> ops) {
+  if (ops.empty())
+    return;
+  auto *ctx = ops[0]->getContext();
+
+  // Ensure entries exist for each operation
+  llvm::for_each(ops, [&](auto *op) { symbolTables.try_emplace(op, nullptr); });
+
+  // Construct them in parallel
+  mlir::parallelForEach(ctx, ops, [&](auto *op) {
+    auto it = symbolTables.find(op);
+    assert(it != symbolTables.end());
+    // Create table if doesn't exist
+    if (!it->second)
+      it->second = ::std::make_unique<InnerSymbolTable>(op);
+  });
+}
+
+//===----------------------------------------------------------------------===//
+// InnerRefNamespace
+//===----------------------------------------------------------------------===//
+
+Operation *InnerRefNamespace::lookup(hw::InnerRefAttr inner) {
+  auto mod = symTable.lookup(inner.getModule());
+  assert(mod->hasTrait<mlir::OpTrait::InnerSymbolTable>());
+  return innerSymTables.getInnerSymbolTable(mod).lookup(inner.getName());
+}
+
+//===----------------------------------------------------------------------===//
+// InnerRef verification
+//===----------------------------------------------------------------------===//
+
+namespace detail {
+
+LogicalResult verifyInnerRefs(Operation *op) {
+  if (op->getNumRegions() != 1)
+    return op->emitOpError() << "Operations with a 'InnerRefNamespace' must "
+                                "have exactly one region";
+  if (!llvm::hasSingleElement(op->getRegion(0)))
+    return op->emitOpError() << "Operations with a 'InnerRefNamespace' must "
+                                "have exactly one block";
+
+  InnerSymbolTableCollection innerSymTables;
+  SymbolTable symbolTable(op);
+  InnerRefNamespace ns{symbolTable, innerSymTables};
+
+  // Gather top-level ops to process in parallel
+  SmallVector<Operation *> childOps(
+      llvm::make_pointer_range(op->getRegion(0).front()));
+
+  // Filter these to those that have the InnerSymbolTable trait,
+  // these will be walked in parallel to find all inner symbols.
+  SmallVector<Operation *> innerSymTableOps(
+      llvm::make_filter_range(childOps, [&](Operation *op) {
+        return op->hasTrait<OpTrait::InnerSymbolTable>();
+      }));
+
+  // Build up the InnerSymbolTables for each op w/InnerSymbolTable trait
+  innerSymTables.constructTablesInParallelFor(innerSymTableOps);
```

Comment:
I'm not crazy about the interface taking an array of ops, can we just have a `InnerSymbolTableCollection::populateTables()` function and it can collect the `InnerSymbolTable` ops itself?

Reply:
Yes, good call.  That is strange.  Also, that way this can be used elsewhere too.  I'll iterate on this a bit, thanks!